cpp/src/arrow/util/bpacking.cc

```diff
@@ -166,14 +169,34 @@ struct Unpack32DynamicFunction {
   }
 };
 
+#if defined(ARROW_HAVE_AVX512_ICX)
+struct Unpack32DynamicFunctionICX {
+  using FunctionType = decltype(&unpack32_default);
+
+  static std::vector<std::pair<DispatchLevel, FunctionType>> implementations() {
+    return {{DispatchLevel::NONE, unpack32_default},
+            {DispatchLevel::AVX512_ICX, unpack32_avx512_icx}};
+  }
+};
```

Comment:
We can't reuse `Unpack32DynamicFunction` for `AVX512_ICX` because `unpack32_avx512_icx()` doesn't support `num_bits > 16`, right?
Can we use `unpack32_specialized()` for `num_bits > 16` like other `unpack32_*()`?

Reply:
Hi, @kou , yes, unpack32_avx512_icx() doesn't support num_bits > 16 currently, but the support of num_bits between 17 to 32 is in progress and almost done. We need to do more tests about this and if it goes well, we will upstream the implementation.

Reply:
Or `unpack32_avx512_icx` should simply fallback on `unpack32_avx512` for `num_bits > 16`. We should not have two different dynamic dispatches for this.

Reply:
@pitrou Hi, we have changed the code per your suggestions. Please review.